### PR TITLE
feat: fetch addresses from chain

### PR DESCRIPTION
**Pull request by fadeev**
see https://github.com/zeta-chain/protocol-contracts/pull/10#issuecomment-1927173560

### CODE REVIEW

**Comment by fadeev** [comment]
Updated to Athens 3. This is what the JSON looks like now. Missing pauser and updated addresses.

```json
{
  "btc_testnet": {
    "chainId": "18332",
    "tss": "tb1qy9pqmk2pd9sv63g27jt8r657wy0d9ueeh0nqur"
  },
  "mumbai_testnet": {
    "chainId": "80001",
    "tss": "0x8531a5aB847ff5B22D855633C25ED1DA3255247e",
    "zrc20ContractAddress": "0x48f80608B672DC30DC7e3dbBd0343c5F02C738Eb",
    "zetaTokenContractAddress": "0x0000c9ec4042283e8139c74f4c64bcd1e0b9b54f",
    "connectorContractAddress": "0x0000ecb8cdd25a18f12daa23f6422e07fbf8b9e1",
    "erc20CustodyContractAddress": "0x0000a7db254145767262c6a81a7ee1650684258e"
  },
  "bsc_testnet": {
    "chainId": "97",
    "tss": "0x8531a5aB847ff5B22D855633C25ED1DA3255247e",
    "zrc20ContractAddress": "0xd97B1de3619ed2c6BEb3860147E30cA8A7dC9891",
    "zetaTokenContractAddress": "0x0000c9ec4042283e8139c74f4c64bcd1e0b9b54f",
    "connectorContractAddress": "0x0000ecb8cdd25a18f12daa23f6422e07fbf8b9e1",
    "erc20CustodyContractAddress": "0x0000a7db254145767262c6a81a7ee1650684258e"
  },
  "goerli_testnet": {
    "chainId": "5",
    "tss": "0x8531a5aB847ff5B22D855633C25ED1DA3255247e",
    "zrc20ContractAddress": "0x13A0c5930C028511Dc02665E7285134B6d11A5f4",
    "zetaTokenContractAddress": "0x0000c304d2934c00db1d51995b9f6996affd17c0",
    "connectorContractAddress": "0x00005e3125aba53c5652f9f0ce1a4cf91d8b15ea",
    "erc20CustodyContractAddress": "0x000047f11c6e42293f433c82473532e869ce4ec5"
  },
  "zeta_testnet": {
    "chainId": "7001",
    "systemContract": "0x91d18e54DAf4F677cB28167158d6dd21F6aB3921",
    "connectorZEVM": "0x239e96c8f17C85c30100AC26F635Ea15f23E9c67",
    "uniswapv2FactoryAddress": "0x9fd96203f7b22bCF72d9DCb40ff98302376cE09c",
    "wZetaContractAddress": "0x5F0b1a82749cb4E2278EC87F8BF6B618dC71a8bf",
    "uniswapv2Router02Address": "0x2ca7d64A7EFE2D62A725E2B35Cf7230D6677FfEe",
    "zetaConnectorZEVMAddress": "0x0000000000000000000000000000000000000000",
    "fungibleModuleAddress": "0x735b14BB79463307AAcBED86DAf3322B1e6226aB"
  }
}
```

**Comment by fadeev** [comment]
If we're ok with this format, I can modify it to generate types as well.

**Comment by fadeev** [comment]
Protocol contract addresses are now generated with the following command:

```
yarn generate
```

On each PR `yarn generate` is executed to check that the contracts are up-to-date.

Addresses can be fetched either by find the contract manually:

```ts
import { mainnet, testnet } from "@zetachain/protocol-contracts";

mainnet.find(
  (a: any) => a.type === "connector" && a.chain_name === "eth_mainnet"
)?.address; // => 0x239e96c8f17C85c30100AC26F635Ea15f23E9c67
```

Or by using `getAddress`:

```ts
import { getAddress } from "@zetachain/protocol-contracts";

getAddress("connector", "zeta_mainnet") // => 0x239e96c8f17C85c30100AC26F635Ea15f23E9c67
```

More examples:

```
getAddress("connector", "eth_mainnet") // => 0x000007Cf399229b2f5A4D043F20E90C9C98B7C6a

getAddress("erc20Custody", "mumbai_testnet") // => 0x0000a7db254145767262c6a81a7ee1650684258e
```

For `zrc20` the function accepts an optional third argument that specifies the symbol:

```
getAddress("zrc20", "zeta_mainnet", "USDT.ETH") // => 0x7c8dDa80bbBE1254a7aACf3219EBe1481c6E01d7
```

The function throws an error when the `type` is not `zrc20`, but the symbol is provided:

```
getAddress("connector", "zeta_mainnet", "USDT.ETH")
Error: Symbol is only supported when ParamType is zrc20
```

This preserves some backwards-compatibility:

```
getAddress("zrc20", "bsc_testnet") // => 0xd97B1de3619ed2c6BEb3860147E30cA8A7dC9891

getAddress("zrc20", "zeta_testnet", "tBNB") // => 0xd97B1de3619ed2c6BEb3860147E30cA8A7dC9891
```

Argument types are generated automatically as well:

Screenshot 2024-02-05 at 18 02 18

**Review by andresaiello — APPROVED** [review]

**Comment by fadeev** [comment]
fwiw this address is not Uniswap, but a ZetaChain connector:

https://github.com/zeta-chain/protocol-contracts/blob/5b7c26d85df77e5487a0ced637680972053b3345/data/addresses.json#L117C29-L117C71

**Comment by fadeev** [comment]
* Added addresses JSONs in the `tasks/` directory, they contain addresses such as ZETA token consumer and uniswap routers which we cannot get from existing contract addresses. If you know how to not have these addresses hard-coded, please, let me know.
* We're now using uniswap router addresses to get WETH and factory addresses. This additionally validates router addresses (wrong router addresses will not have WETH and factory getters) and allows us to not have WETH and factory hard-coded.

**Review by andresaiello — DISMISSED** [review]

File: data/addresses.mainnet.json [hunk 1]
```diff
@@ -0,0 +1,289 @@
+[
+  {
+    "address": "0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f",
+    "category": "messaging",
```
Comment (lumtis) [inline data/addresses.mainnet.json:4]: Uniswap pool can be used for omnichain smart contract call as well if a ERC20 deposit reverts.

I think we should have a category: general/misc/common

**Review by lumtis — APPROVED** [review]

File: data/addresses.mainnet.json (hunk 1 above)
Comment (fadeev) [reply data/addresses.mainnet.json:4]: You mean if a smart contract on a connected chain deposits an ERC-20 token to ZetaChain, this reverts, the contract can then use the Uniswap address to swap the token and return it to the user?

File: lib/address.tools.ts [hunk 2]
```diff
@@ -81,13 +81,13 @@ export const isMainnetNetwork = (network: ZetaProtocolTestNetwork): boolean => {
   return false;
 };
 
-export const getAddress = (address: ZetaProtocolAddress | ZetaZEVMAddress, network: ZetaProtocolNetwork): string => {
-  if (isZetaProtocolAddress(address)) {
-    return (addresses["ccm"] as any)[network][address];
-  }
+// export const getAddress = (address: ZetaProtocolAddress | ZetaZEVMAddress, network: ZetaProtocolNetwork): string => {
```
Comment (andresaiello) [inline lib/address.tools.ts:84]: any new version of this?

**Review by andresaiello — APPROVED** [review]

**Pull request merged**